### PR TITLE
feat: validate input on run

### LINKED ISSUE #66: Add INPUT schema checks to "apify run"
If actor has an INPUT schema, then "apify run" should validate the input against the schema, using the exactly same code as Apify cloud is doing that.

### PR DESCRIPTION
**Pull request by vladfrangu**
Closes #66 

Still waiting on https://apify.slack.com/archives/G0134LE0JUB/p1714484758429109

### CODE REVIEW

**Review by B4nan — COMMENTED** [review]
some tests would be nice, otherwise looking good

**Comment by jancurn** [comment]
Thanks @vladfrangu 

@netmilk please did you review this is consistent with the new [CLI design](https://www.notion.so/apify/API-CLI-design-Final-Draft-517b46afd5b8431dae7e38b06c9080fc)? 

Guys, please assign Adam as reviewers to all PRs related to CLI, to ensure consistency with the new design

File: src/commands/run.ts [hunk 1]
```diff
@@ -174,103 +195,244 @@ export class RunCommand extends ApifyCommand<typeof RunCommand> {
...
             }
+        } finally {
+            // Delete the temporary input file (but maybe we should keep it?)
+            if (envVariableOverride) {
+                await deleteFile(join(process.cwd(), getLocalKeyValueStorePath(), `${envVariableOverride}.json`));
+            }
+        }
+    }
+
+    private async validateAndStoreInput(inputOverride?: Record<string, unknown>) {
```
Comment (jancurn) [inline src/commands/run.ts:358]: Comment would be good

File: src/lib/input_schema.ts [hunk 2]
```diff
@@ -102,3 +103,42 @@ export const createPrefilledInputFileFromInputSchema = async (actorFolderDir: st
         await writeJsonFile(inputJsonPath, inputFile);
     }
 };
+
+export const getDefaultsAndPrefillsFromInputSchema = (inputSchema: any) => {
```
Comment (jancurn) [inline src/lib/input_schema.ts:107]: Comments please, especially for exported functions

Comment (vladfrangu) [reply src/lib/input_schema.ts:107]: I'll add these in a follow up PR

**Comment by netmilk** [comment]
@jancurn Yes, I've validated it in the E2E product vise. The flags interface design follows the new design style guide. But I've just noticed that piping to `stidn` without a flag as an alias to `--input-file=-` seems undocumented. Is it @vladfrangu, or am I missing it somewhere?

```
$ jo usernames[]=netmilk | apify call netmilk/sherlock
Run: Calling Actor netmilk/sherlock (sSuwVmyjxzabXcoMh)
2024-07-01T15:26:11.619Z ACTOR: Pulling Docker image of build ASIrG38fzQ3FLo0YO from repository.
2024-07-01T15:26:26.570Z ACTOR: Creating Docker container.
2024-07-01T15:26:26.740Z ACTOR: Starting Docker container.
...
```

```
$ apify call --help
...
FLAGS
...
  -i, --input=<value>            Optional JSON input to be given to the Actor.
...
      --input-file=<value>       Optional path to a file with JSON input to be given to the Actor. The file must be a valid JSON file. You can
                                 also specify `-` to read from standard input.
```

However, the canonical `call` command itself is supposed to be under the `$ apify actors ` namespace in the future CLI design, and the `call` on root is going to be an alias to that.

Note to self: That reminds me we should create a milestone and a plan for the transformation to the "new CLI design" and make the design a public, living document here in this repo.

**Comment by vladfrangu** [comment]
This PR was before our discussion about stdin flags @netmilk, I will fix it today in a subsequent release. Also this is about the `run` command, not the `call` command

**Comment by netmilk** [comment]
@vladfrangu @jancurn Darn, my bad. 🤦‍♂️ Silly. Apologies. I'll look into it.

**Comment by netmilk** [comment]
@vladfrangu I need help figuring out whether this is even changing the interface. I assume it doesn't. But apart from that, it works as expected. Thank you!

Shall we consider returning the JSON path (pointer) to the INPUT data structure in the error message in case it isn't flat? It's going to be hard to figure out which part of the input is failing the validation otherwise.

```
$ jo foo=bar | apify run
    Error: The input in your storage is invalid. Please fix the following errors:
      - Field awesome is required
$ echo $?
1
```
```
$ apify run --input=$(jo foo=bar)
    Error: The input provided is invalid. Please fix the following errors:
      - Field awesome is required
```
```
$ jo foo=bar > invalid.json && apify run --input-file=invalid.json
    Error: The input provided is invalid. Please fix the following errors:
      - Field awesome is required
```      
```
$ apify run --input-file=invalid.json 2> error | cat -
$ cat invalid.json | apify run --input-file=-
Error: Cannot read input file at path "/Users/netmilk/projects/apify-cli-validate/test-input-validation-actor/-".
  ENOENT: no such file or directory, open '/Users/netmilk/projects/apify-cli-validate/test-input-validation-actor/-'
$ echo $?
5
```
```
$ apify run --input-file=invalid.json 2> error
$ cat error
    Error: The input provided is invalid. Please fix the following errors:
      - Field awesome is required
```

I've noticed one minor thing, though. The `--input-file=-' doesn't work the same way it does in the `call` command, can you fix that for the sake of consistency, please?

**Comment by vladfrangu** [comment]
Yep, that's the plan for today, could you maybe also make an issue for this so we have it fully tracked? 🙏

**Pull request merged**